### PR TITLE
make query detect stop byte

### PR DESCRIPTION
**Pull request by chadek**
This PR should allow to autodetect the stop byte and get reed of all the buffersize stuff. This a for instance a draft (so DO NOT MERGE for now) that I still need to test on my hardware. If it does work, I will also delete all the buffer size related stuff which will make everything simpler

### CODE REVIEW

**Comment by A13xund3r** [comment]
Does it compile for you?

```
bananapi:skymax-demo:# make                                                                                    <master ✗>
[ 20%] Building CXX object CMakeFiles/inverter_poller.dir/inverter.cpp.o
/root/github/skymax-demo/inverter.cpp: In member function ‘bool cInverter::query(const char*, int)’:
/root/github/skymax-demo/inverter.cpp:116:11: error: ‘READ_BUFFER_SIZE’ does not name a type
  116 |     const READ_BUFFER_SIZE = 60;
      |           ^~~~~~~~~~~~~~~~
/root/github/skymax-demo/inverter.cpp:120:32: warning: pointer of type ‘void *’ used in arithmetic [-Wpointer-arith]
  120 |         n = read(fd, (void*)buf+i, READ_BUFFER_SIZE);
      |                      ~~~~~~~~~~^~
/root/github/skymax-demo/inverter.cpp:120:36: error: ‘READ_BUFFER_SIZE’ was not declared in this scope
  120 |         n = read(fd, (void*)buf+i, READ_BUFFER_SIZE);
      |                                    ^~~~~~~~~~~~~~~~
/root/github/skymax-demo/inverter.cpp:158:60: error: ‘j’ was not declared in this scope
  158 |         lprintf("INVERTER: %s reply size (%d bytes)", cmd, j);
      |                                                            ^
make[2]: *** [CMakeFiles/inverter_poller.dir/build.make:90: CMakeFiles/inverter_poller.dir/inverter.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/inverter_poller.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

**Comment by chadek** [comment]
No it didn't, now it does

**Comment by A13xund3r** [comment]
Now my life is much easier.. but there is strange behaviour in one case. 
TLDR:
If transmission is incomplete (some data at end is not received including stop and CRC bytes I'm getting 
```
ananapi:skymax-demo:# ./inverter_poller -d -r QBMS                                                                                          <master ✗>
Thu Nov  2 11:47:10 2023 INVERTER: Debug set
Thu Nov  2 11:47:10 2023 INVERTER: Current CRC: 61 44
Thu Nov  2 11:47:10 2023 INVERTER: QBMS reply size (32 bytes)
Thu Nov  2 11:47:10 2023 INVERTER: QBMS: CRC Failed!  Reply size: 32  Buffer: (1 000 0 0 0 000 000 000 0000 00r the actual inverter polling process...

```
So it still consider that transmission is completed.
Here I need explain my circumstance.
For unknown (yet) reason sometimes I'm getting NAK response to valid command, then it will respond this way to every command regardless it is valid or not. Until I run serial port sniffer on another console, then it will do NAK once again and return to normal replies. Problem is that sniffer I found and use is quite old and taking over serial port access in random way, depend which program is first when data arrive something is missing in inverter poller or sniffer gets `Error reading from port: Resource temporarily unavailable`. So for case above sniffer got remaining part of message `Device --> 0 (048) 0 (048) � (128) � (134) <CR> (013)` there are 2 bytes of message body, CRC and CR, when all data is complete it is like that:
```
bananapi:skymax-demo:# ./inverter_poller -d -r QBMS                                                               
Thu Nov  2 11:47:25 2023 INVERTER: Debug set
Thu Nov  2 11:47:25 2023 INVERTER: Current CRC: 61 44
Thu Nov  2 11:47:25 2023 INVERTER: QBMS reply size (37 bytes)
Thu Nov  2 11:47:25 2023 INVERTER: QBMS: 37 bytes read: (1 000 0 0 0 000 000 000 0000 0000
Thu Nov  2 11:47:25 2023 INVERTER: QBMS query finished
Reply:  1 000 0 0 0 000 000 000 0000 0000
```

**Comment by chadek** [comment]
I'm glad to ear that it seams working. I'll get my hands on my hardware sunday or monday to test this out.

Again, this is still a draft, I still need to make some adjustments as I choose to "stream" the file descriptor loading a certain quantity of byte (60) which should work for most of the small commands on the first batch. I see some things that are probably not good but without anything to test for now I can do much.

**Comment by A13xund3r** [comment]
Does `const int READ_BUFFER_SIZE = 60;` limit RX buffer to 60 bytes? If yes then it is too short, my longest reply from inverter is 143 bytes.

**Comment by chadek** [comment]
I tried on my hardware, raw command seams to work properly but now the use of the tool normally without options seams to randomly work and printing crazy values on some variables probably due to a wrong buffer size badly detected for some reason. I'll dig more when I'll have time, probably this coming week

**Comment by A13xund3r** [comment]
I don't know how to updated the code with your changes other way than just copy paste changes and deleting stuff manually, or maybe @chadek your fork does include all changes? Can someone point me in right direction?

**Comment by chadek** [comment]
You can clone my repo and checkout on query_rework branch (git checkout query_rework)

**Comment by A13xund3r** [comment]
> @A13xund3r, I fixed the query function to detect stop byte, it should detect automatically the reply length. Every raw cmd should work and you should be able to read data without any replysize config which I removed from code as it should be now useless. Tell me if it works on your hardware

So using your fork and query_rework branch it seems that CR detection is not working properly
```
bananapi:skymax-demo:# ./inverter_poller -l 16 -d -r QT
Thu Nov 16 12:04:40 2023 INVERTER: Debug set
Thu Nov 16 12:04:40 2023 INVERTER: Current CRC: 27 FF
Thu Nov 16 12:04:43 2023 INVERTER: QT read timeout
Thu Nov 16 12:04:43 2023 INVERTER: QT command timeout, or couldn't find stop byte. Byte read (18 bytes)
Reply:  

```
when sniffer proves it is there
`Device --> ( (040) 2 (050) 0 (048) 2 (050) 3 (051) 1 (049) 1 (049) 2 (050) 4 (052) 0 (048) 4 (052) 3 (051) 7 (055) 4 (052) 5 (053) � (138) � (221) <CR> (013)`
Also reply is empty when I expect it to contain received data despite it is assumed incomplete.

**Comment by chadek** [comment]
2 possibility, your command really timeout and you probably have some weird communication issue I also have sometimes maybe because of the cable not plugged well (unplugged replug) or your inverter was still trying to send data when command timeout was reached. In that case, try to put a higher value than 2 (seconds) on inverter.cpp line 119 to see what happen. Does other commands work?

**Comment by A13xund3r** [comment]
with j<=i+n, both usleeps=10, READ_BUFFER_SIZE is 15 so I don't touch it for now
```
bananapi:skymax-demo:# ./inverter_poller -d -r QT
Fri Nov 17 13:46:50 2023 INVERTER: Debug set
Fri Nov 17 13:46:51 2023 INVERTER: Current CRC: 27 FF
Fri Nov 17 13:46:58 2023 INVERTER: QT read timeout
Fri Nov 17 13:46:58 2023 INVERTER: QT command timeout, or couldn't find stop byte. Byte read (7 bytes). Buffer: (NAKss
****deleted content of config file****
Reply:  
```
Please note it got 7 bytes, start byte(1), NAK(3),CRC(2), stop byte(1). so what it got as last byte? It seems to be not printable character.
Started sniffer, then:
```
bananapi:skymax-demo:# ./inverter_poller -d -r QT
Fri Nov 17 13:49:50 2023 INVERTER: Debug set
Fri Nov 17 13:49:50 2023 INVERTER: Current CRC: 27 FF
INVERTER: stop byte detected, buffersize might be 7 for QT Fri Nov 17 13:49:50 2023 INVERTER: QT reply size (7 bytes)
Fri Nov 17 13:49:50 2023 INVERTER: QT: 7 bytes read: (NAK 
Fri Nov 17 13:49:50 2023 INVERTER: QT query finished
Reply:  NAK
bananapi:skymax-demo:# ./inverter_poller -d -r QT
Fri Nov 17 13:50:27 2023 INVERTER: Debug set
Fri Nov 17 13:50:27 2023 INVERTER: Current CRC: 27 FF
Fri Nov 17 13:50:34 2023 INVERTER: QT read timeout
ic configuration options for the actual inverter polling process...find stop byte. Byte read (0 bytes). Buffer: QT'�
Reply:  
```
second reply was blocked by sniffer which got: `Device --> ( (040) 2 (050) 0 (048) 1 (049) 8 (056) 0 (048) 2 (050) 2 (050) 4 (052) 0 (048) 6 (054) 2 (050) 9 (057) 1 (049) 5 (053) � (193) � (244) <CR> (013)`
then 3-rd attempt:
```
bananapi:skymax-demo:# ./inverter_poller -d -r QT
Fri Nov 17 13:50:37 2023 INVERTER: Debug set
Fri Nov 17 13:50:37 2023 INVERTER: Current CRC: 27 FF
INVERTER: stop byte detected, buffersize might be 18 for QT Fri Nov 17 13:50:37 2023 INVERTER: QT reply size (18 bytes)
Fri Nov 17 13:50:37 2023 INVERTER: QT: 18 bytes read: (20180224062925 
Fri Nov 17 13:50:37 2023 INVERTER: QT query finished
Reply:  20180224062925
```
Killed sniffer then:
```
bananapi:skymax-demo:# ./inverter_poller -d -r QT
Fri Nov 17 14:19:13 2023 INVERTER: Debug set
Fri Nov 17 14:19:13 2023 INVERTER: Current CRC: 27 FF
Fri Nov 17 14:19:20 2023 INVERTER: QT read timeout
Fri Nov 17 14:19:20 2023 INVERTER: QT command timeout, or couldn't find stop byte. Byte read (18 bytes). Buffer: (20180224065802�?
bananapi:skymax-demo:# ./inverter_poller -d -r QT  
Fri Nov 17 14:25:14 2023 INVERTER: Debug set
Fri Nov 17 14:25:14 2023 INVERTER: Current CRC: 27 FF
Fri Nov 17 14:25:21 2023 INVERTER: QT read timeout
Fri Nov 17 14:25:21 2023 INVERTER: QT command timeout, or couldn't find stop byte. Byte read (7 bytes). Buffer: (NAKss
```
I will try to investigate flow control configuration then will make hardware serial port sniffer to see what is going on wires. Fact that communication is pretty reliable with sniffer active suggests there is port config issue. Alternatively I'll wait until another banana pi I ordered arrive and connect it to USB port, this banana is using almost 30m cable so it is too far for USB.

**Comment by chadek** [comment]
This looks like indeed a weird communication issue if it works sometimes and sometimes not

**Comment by A13xund3r** [comment]
Got it... following https://blog.mbedded.ninja/programming/operating-systems/linux/linux-serial-ports-using-c-cpp/ I've added: 
`settings.c_iflag &= ~(IGNBRK|BRKINT|PARMRK|ISTRIP|INLCR|IGNCR|ICRNL); // Disable any special handling of received bytes`
now CR is properly recognized every time, also I added:
```
    settings.c_lflag &= ~ICANON;
    settings.c_iflag &= ~(IXON | IXOFF | IXANY); // Turn off s/w flow ctrl
```
line 86 `cfsetospeed(&settings, baud);      // baud rate` changed to `cfsetspeed(&settings, baud);      // baud rate`
first one set baud for output stream only when latter for both directions. With this everything is working fine so far.

**Comment by chadek** [comment]
Thanks for the digging, now it work everytime with this new parameters! Could you check if this last version work on your side? Then I'll set this PR reading to merge :)

**Comment by A13xund3r** [comment]
In my case outgoing stream seems to be corrupted, I have to run sniffer to get transfers without issues or after first command I'm getting NAKs, LF is added? I suspect `settings.c_oflag &= ~OPOST;        // Prevent special interpretation of output bytes (e.g. newline chars)` is not enough. Will dig deeper.
```
bananapi:skymax-demo:# ./inverter_poller -r QT 
INVERTER: stop byte detected, buffersize might be 18 for QT Reply:  20231121135659
bananapi:skymax-demo:# ./inverter_poller -r QT 
INVERTER: stop byte detected, buffersize might be 7 for QT Reply:  NAK
bananapi:skymax-demo:# ./inverter_poller -r QT 
INVERTER: stop byte detected, buffersize might be 7 for QT Reply:  NAK
```

**Comment by A13xund3r** [comment]
Making separate message because previous post edit apparently is not notified.

**Comment by chadek** [comment]
Sorry for the delay, I added these new parameters. I haven't my hardware on hands to test for now but if it is working for you, we finally could mark this pr ready to merge

**Comment by chadek** [comment]
Did not tried my last code yet, maybe today and then mark this PR as ready to merge. In my case, I choose to use this project instead : https://github.com/jblance/mpp-solar because it support a large amount of devices and seams to be maintained. The only thing I do not really like is that is it is quite complex, and by consequence bit heavier on hardware :'c

**Comment by manio** [comment]
Thanks!

**Comment by synogen** [comment]
> Did not tried my last code yet, maybe today and then mark this PR as ready to merge. In my case, I choose to use this project instead : https://github.com/jblance/mpp-solar because it support a large amount of devices and seams to be maintained. The only thing I do not really like is that is it is quite complex, and by consequence bit heavier on hardware :'c

I just checked their wiki in more detail and I hadn't realized until now they had MQTT support and can even format the messages for Home Assistant which is what I was after. I was aware of the project before but only used it for their nice collection of protocol documents, I only skimmed their readme before this and since I found no mention of MQTT or Home Assistant there I didn't delve any deeper.

Thanks for pointing me to it again, I would've completely missed all the things that one can do otherwise! 👍
